### PR TITLE
Add query timeout Support

### LINKED ISSUE #146: Support query timeout
The original promql engine supports specifying query timeout https://github.com/prometheus/prometheus/blob/main/promql/engine.go#L249.

It would be nice to have that feature in this engine as well.

### PR DESCRIPTION
**Pull request by PradyumnaKrishna**
Fixes #146 

Current engine doesn't support query timeout, which was a part of original promql-engine.
This Pull Request adds timeout option in `compatibilityEngine`, as our engine can't access
private data member `timeout` of the `promql.Engine`.

Testcase are updated as well.

### CODE REVIEW

File: engine/engine.go [hunk 1]
```diff
@@ -145,6 +146,7 @@ func New(opts Opts) *compatibilityEngine {
 
 type compatibilityEngine struct {
 	prom    *promql.Engine
+	timeout time.Duration
```
Comment (fpetkovski) [inline engine/engine.go:149]: Let's maybe move the timeout bellow the `logicalOptimizers` variable to group query parameters in one place?

File: engine/engine_test.go [hunk 2]
```diff
@@ -2538,6 +2538,34 @@ func TestQueryCancellation(t *testing.T) {
 	testutil.Equals(t, context.Canceled, newResult.Err)
 }
 
+func TestQueryTimeout(t *testing.T) {
+
```
Comment (fpetkovski) [inline engine/engine_test.go:2542]:
```suggestion
```

File: engine/engine_test.go [hunk 3]
```diff
@@ -2538,6 +2538,34 @@ func TestQueryCancellation(t *testing.T) {
...
+	end := time.Unix(120, 0)
+
```
Comment (fpetkovski) [inline engine/engine_test.go:2544]:
```suggestion
```

**Review by fpetkovski — COMMENTED** [review]
This looks great, mostly code style comments from my side. 

Just to confirm, will all queries fail when timeout is set to 0? And is this how the Prometheus engine behaves as well?

**Comment by PradyumnaKrishna** [comment]
> Just to confirm, will all queries fail when timeout is set to 0? And is this how the Prometheus engine behaves as well?

Yes, I have confirmed this by creating some testcase with Prometheus engine as well.

**Review by fpetkovski — APPROVED** [review]
Thanks, LGTM!

**Pull request merged**